bundles/org.openhab.binding.radiothermostat/src/main/java/org/openhab/binding/radiothermostat/internal/RadioThermostatHandlerFactory.java

```diff
@@ -0,0 +1,108 @@
+
+import static org.openhab.binding.radiothermostat.internal.RadioThermostatBindingConstants.THING_TYPE_RTHERM;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Hashtable;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerFactory;
+import org.eclipse.smarthome.io.net.http.HttpClientFactory;
+import org.openhab.binding.radiothermostat.internal.discovery.RadioThermostatDiscoveryService;
+import org.openhab.binding.radiothermostat.internal.handler.RadioThermostatHandler;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * The {@link RadioThermostatHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author Michael Lobstein - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = ThingHandlerFactory.class, configurationPid = "binding.radiothermostat")
+public class RadioThermostatHandlerFactory extends BaseThingHandlerFactory {
+
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections.singleton(THING_TYPE_RTHERM);
+    private @Nullable RadioThermostatStateDescriptionProvider stateDescriptionProvider;
+    private final Map<ThingUID, ServiceRegistration<?>> discoveryServiceRegs = new HashMap<>();
+    private final HttpClient httpClient;
+
+    @Activate
+    public RadioThermostatHandlerFactory(final @Reference HttpClientFactory httpClientFactory) {
+        this.httpClient = httpClientFactory.getCommonHttpClient();
+    }
+
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        return SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
+    }
+
+    @Override
+    protected @Nullable ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (thingTypeUID.equals(THING_TYPE_RTHERM)) {
+            RadioThermostatHandler handler = new RadioThermostatHandler(thing, stateDescriptionProvider, httpClient);
+
+            RadioThermostatDiscoveryService discoveryService = new RadioThermostatDiscoveryService(httpClient);
+            discoveryServiceRegs.put(handler.getThing().getUID(), bundleContext
+                    .registerService(DiscoveryService.class.getName(), discoveryService, new Hashtable<>()));
```

Comment:
It doesn't appear as if the discovery service is in any way tied to the ThingHandler instance. I also don't think there should be multiple discovery services either, so you should just make your RadioThermostatDiscoveryService a osgi service instead of initialize it here.

Please add this above your RadioThermostatDiscoveryService class:
```
@Component(immediate=true, service = DiscoveryService.class, configurationPid = "discovery.radiothermostat")
```

then remove all RadioThermostatDiscoveryService and ServiceRegistration logic from the RadioThrmostatHandlerFactory.

Reply:
I did based on example that I found that passed the httpClient from the HandlerFactory into the Discovery service. Is there a better way to pass httpClient?

Reply:
My comments below will show exactly how to get an HttpClient into your DiscoveryService. It is the same way that you inject an HttpClient into this class.

Reply:
Actually, per latest comments, you don't need httpClient in your discovery service at all.

Reply:
yep. I removed it completely.